### PR TITLE
[gfx] metal backend skeleton

### PR DESCRIPTION
**Pull request by westlicht**
This is based on #4220 with some cleanup:
- `metal-cpp` is added as a git submodule
- removed `metal-cpp-extensions`
- cleanup CMake files and make sure the metal backend skeleton compiles
- return `SLANG_E_NOT_IMPLEMENTED` in unimplemented stubs

### CODE REVIEW

File: tools/gfx/metal/metal-swap-chain.cpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        //    assert(drawable);
+        // }
+        // m_drawables[i] = drawable;
+        //MTL::Texture* tex = drawable->texture();
+
+        ITextureResource::Desc imageDesc = {};
+        imageDesc.allowedStates = ResourceStateSet(
+            ResourceState::Present, ResourceState::RenderTarget, ResourceState::CopyDestination);
+        imageDesc.type = IResource::Type::Texture2D;
+        imageDesc.arraySize = 0;
```
Comment (jkwak-work) [inline tools/gfx/metal/metal-swap-chain.cpp:52]: Feel free to ignore this comment but `arraySize` may need to be 1.

File: tools/gfx/metal/metal-device.cpp [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+static bool shouldDumpPipeline()
+{
+    StringBuilder dumpPipelineSettings;
+    PlatformUtil::getEnvironmentVariable(toSlice("SLANG_GFX_DUMP_PIPELINE"), dumpPipelineSettings);
+    return dumpPipelineSettings.produceString() == "1";
+}
+
+DeviceImpl::~DeviceImpl()
+{
+}
```
Comment (jkwak-work) [inline tools/gfx/metal/metal-device.cpp:42]: I understand that the code is incomplete yet.
But it looks like you need to stop the capture manager here as well.

File: tools/gfx/metal/metal-command-encoder.cpp [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+    static const int kMaxScissorRects = 9; // TODO:
+    assert(count < kMaxScissorRects);
+
+    m_scissorRects.setCount(count);
+    for (GfxIndex i = 0; i < count; ++i)
+    {
+        const auto& inRect = rects[i];
+        auto& metalRect = m_scissorRects[i];
+        metalRect.height = inRect.maxX - inRect.minX;
+        metalRect.width = inRect.maxY - inRect.minY;
```
Comment (jkwak-work) [inline tools/gfx/metal/metal-command-encoder.cpp:229]: I am not sure if this is correct.
Y is usually for height.

**Review by jkwak-work — APPROVED** [review]
Looks good to me.

**Comment by skallweitNV** [comment]
@jkwak-work thanks for looking at this. I've already done some follow-up work in a local branch and will put up additional PRs for improving/implementing parts of the backend.

**Pull request merged**